core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java

```diff
@@ -0,0 +1,204 @@
+  private final Closer mCloser;
+  private final BlockWorkerClient mBlockWorkerClient;
+  private final PacketOutStream mOutStream;
+  private boolean mClosed;
+
+  /**
+   * Creates a new local block output stream.
+   *
+   * @param blockId the block id
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if an I/O error occurs
+   * @return the {@link BlockOutStream} instance created
+   */
+  public static BlockOutStream createLocalBlockOutStream(long blockId,
+      long blockSize,
+      WorkerNetAddress workerNetAddress,
+      BlockStoreContext context,
+      OutStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      BlockWorkerClient client = closer.register(context.createWorkerClient(workerNetAddress));
+      PacketOutStream outStream =
+          PacketOutStream.createLocalPacketOutStream(client, blockId, blockSize);
+      closer.register(outStream);
+      return new BlockOutStream(outStream, blockId, blockSize, client, options);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  /**
+   * Creates a new remote block output stream.
+   *
+   * @param blockId the block id
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if an I/O error occurs
+   * @return the {@link BlockOutStream} instance created
+   */
+  public static BlockOutStream createRemoteBlockOutStream(long blockId,
+      long blockSize,
+      WorkerNetAddress workerNetAddress,
+      BlockStoreContext context,
+      OutStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      BlockWorkerClient client = closer.register(context.createWorkerClient(workerNetAddress));
+
+      PacketOutStream outStream = PacketOutStream.createNettyPacketOutStream(
+          client.getDataServerAddress(), client.getSessionId(), blockId, blockSize,
+          Protocol.RequestType.ALLUXIO_BLOCK);
+      closer.register(outStream);
+      return new BlockOutStream(outStream, blockId, blockSize, client, options);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  // Explicitly overriding some write methods which are not efficiently implemented in
+  // FilterOutStream.
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    mOutStream.write(b);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    mOutStream.write(b, off, len);
+  }
+
+  @Override
+  public long remaining() {
+    return mOutStream.remaining();
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    Throwable throwable = null;
+    try {
+      mOutStream.cancel();
```

Comment:
We should probably in the future have a class `Canceler` that cancels all `Cancelable` like `Closer`